### ISSUE TITLE
Authentication fails with arbiter in cluster

### ISSUE THREAD

**Issue by xnagpa** [elixir-mongo/mongodb#135]
Hello! 

Recently I've seen that you updated the driver up to 0.4.0.  So I decided to update my code. 
And at the moment my code can't get the mongo authentication.  I doubt that I do everything correctly, though.

When I downgrade mongodb to v0.2.1 to this commit 4c174516599388bc6ab61bd6ea88dbcef4e14520, everything becomes fine.

I wonder what could be the cause of the problem.

Here are my credentials, that don't work anymore: 

```
worker(Mongo, [[name: :mongo, 
hostname: Application.get_env(:my_app, :mongo)[:hostname], 
port: Application.get_env(:my_app, :mongo)[:port], 
username: Application.get_env(:my_app, :mongo)[:username],
 password: Application.get_env(:my_app, :mongo)[:password], 
database: Application.get_env(:my_app, :mongo)[:database], 
auth_source: Application.get_env(:my_app, :mongo)[:auth_source],
 pool: DBConnection.Poolboy]])
```

Actually, I know that my dbs are arranged into replica set. So I decided to try new feature with seeds. 

```
worker(Mongo, [[name: :mongo, 
seeds: ["my_mongo_01:27017", "my_mongo_02:27017"]
username: Application.get_env(:my_app, :mongo)[:username],
 password: Application.get_env(:my_app, :mongo)[:password], 
database: Application.get_env(:my_app, :mongo)[:database], 
auth_source: Application.get_env(:my_app, :mongo)[:auth_source],
 pool: DBConnection.Poolboy]])
```
It didn't succeed either. This is what I got in both cases after upgrade:

> 15:35:15.348 [error] Mongo.Protocol (#PID<0.1487.0>) failed to connect: ** (Mongo.Error) auth failed for user myUser: Authentication failed. 18
> 15:35:15.411 [error] Mongo.Protocol (#PID<0.1484.0>) failed to connect: ** (Mongo.Error) auth failed for user myUser: Authentication failed. 18
> 15:35:15.697 [error] Mongo.Protocol (#PID<0.1489.0>) failed to connect: ** (Mongo.Error) auth failed for user myUser: Authentication failed. 18
> 15:35:15.754 [error] Mongo.Protocol (#PID<0.1490.0>) failed to connect: ** (Mongo.Error) auth failed for user myUser: Authentication failed. 18
> 15:35:15.845 [error] Mongo.Protocol (#PID<0.1493.0>) failed to connect: ** (Mongo.Error) auth failed for user myUser: Authentication failed. 18
> 15:35:15.867 [error] Mongo.Protocol (#PID<0.1491.0>) failed to connect: ** (Mongo.Error) auth failed for user myUser: Authentication failed. 18

**Comment by ankhers** [comment]
Can you give some more information about how this Mongo cluster is setup?

**Comment by xnagpa** [comment]
My app calls mongo from another server. 
Mongo cluster is a simple replica set, which consists of two replicas and arbiter. Replicas receive requests on 27017 port, and arbiter receives'em on 27018.  
When I connect to mongo via 0.4.0 driver I specify replicas addresses(arbiter is not specified). 
I also tried to connect to just one replica, using 'host' parameter. It was okay on 0.3.0, but doesn't work anymore. 

Please tell me if this information is enough.

**Comment by xnagpa** [comment]
The issue doesn't depend on fact whether I specified arbiter's address or not.  Looks like it tries to find my user in the auth_database and doesn't find it .

**Comment by xnagpa** [comment]
So after a little research I found out, that driver tries to authenticate against arbiter node. Which causes an instant connection fail.

From my arbiter's log

```
2017-06-26T15:46:16.518+0300 I NETWORK  [initandlisten] connection accepted from 10.41.5.27:40928 #4052 (335 connections now open)
2017-06-26T15:46:16.519+0300 I ACCESS   [conn4052] SCRAM-SHA-1 authentication failed for MyUser on admin from client 10.44.5.27 ; UserNotFound: Could not find user MyUser@admin
```

I found issue in node's mongo driver which looks exactly same as mine: 
https://jira.mongodb.org/browse/NODE-981

**Comment by ankhers** [comment]
Thanks for finding the root of the problem. I will look into getting a fix out.

**Comment by svsdehh** [comment]
we are having the same problem. Do you have any new status for this issue?

**Comment by ankhers** [comment]
I can't seem to replicate this issue locally. Do either of you have a cluster that I would be able to test on where this happens?

**Comment by svsdehh** [comment]
thanks for you afford to fix this.

it happens in our company where the cluster is not accessible from the outside. I also only get:

[error] Mongo.Protocol (#PID<0.297.0>) failed to connect: ** (Mongo.Error) auth failed for user XXX: Authentication failed. 18

if you have a hint, where this happens, I can try to see, whether I can debug this myself a little bit.

**Comment by ankhers** [comment]
Unfortunately I can only guess at where the issue is coming from. However, I would take a look at `Mongo.Monitor.check/1`. That is where we are asking what the `server_description` for a given server looks like.

**Comment by ankhers** [comment]
Have either of you made any headway with this issue?

**Comment by svsdehh** [comment]
Hi!

we got some infos that might help you to replicate the problem.
First our setup:

We have 
ONE primary (db-dev-706784.company.com:27017)
ONE secondary (db-dev-706783.company.com:27017)
ONE arbiter (db-dev-706875.company.com:27017)

These are the logs we get for the `server_description` inside the `check/1` method:

```
%{address: "db-dev-706783.company.com:27017",
  arbiters: ["db-dev-706875.company.com:27017"], election_id: nil,
  error: nil,
  hosts: ["db-dev-706783.company.com:27017",
   "db-dev-706784.company.com:27017"], last_update_time: -576460449584,
  last_write_date: #<DateTime(2017-07-25T08:25:42.000Z Etc/UTC)>,
  max_wire_version: 5, me: "db-dev-706783.company.com:27017",
  min_wire_version: 0,
  op_time: %{"t" => -1, "ts" => #BSON.Timestamp<6446621967129772037>},
  passives: [], primary: "db-dev-706784.company.com:27017",
  round_trip_time: 3, set_name: "develop_RS", set_version: 597,
  tag_set: %{"dc" => "KGHAM"}, type: :rs_secondary}
%{address: "db-dev-706784.company.com:27017",
  arbiters: ["db-dev-706875.company.com:27017"],
  election_id: #BSON.ObjectId<5976a54ccfdebd42b8d1cc6c>, error: nil,
  hosts: ["db-dev-706783.company.com:27017",
   "db-dev-706784.company.com:27017"], last_update_time: -576460449584,
  last_write_date: #<DateTime(2017-07-25T08:25:42.000Z Etc/UTC)>,
  max_wire_version: 5, me: "db-dev-706784.company.com:27017",
  min_wire_version: 0,
  op_time: %{"t" => -1, "ts" => #BSON.Timestamp<6446621967129772037>},
  passives: [], primary: "db-dev-706784.company.com:27017",
  round_trip_time: 3, set_name: "develop_RS", set_version: 597,
  tag_set: %{"dc" => "KGHAM"}, type: :rs_primary}
%{address: "db-dev-706875.company.com:27017",
  arbiters: ["db-dev-706875.company.com:27017"], election_id: nil,
  error: nil,
  hosts: ["db-dev-706783.company.com:27017",
   "db-dev-706784.company.com:27017"], last_update_time: -576460449784,
  last_write_date: #<DateTime(2017-07-25T08:25:42.000Z Etc/UTC)>,
  max_wire_version: 5, me: "db-dev-706875.company.com:27017",
  min_wire_version: 0,
  op_time: %{"t" => -1, "ts" => #BSON.Timestamp<6446621967129772037>},
  passives: [], primary: "db-dev-706784.company.com:27017",
  round_trip_time: 3, set_name: "develop_RS", set_version: 597,
  tag_set: %{}, type: :rs_arbiter}

```


It seems the auth call is made for each of these three servers incl. the arbiter. Which then leads into an error:

```
{:ok,
 %{"code" => 18, "codeName" => "AuthenticationFailed",
   "errmsg" => "Authentication failed.", "ok" => 0.0}}
```


The application is able to start and run but we get priodically every few seconds an auth error in the log file.


Even if we dont get the same error code we can reproduce this while connecting to the mongo nodes directly:


```
~$ mongo -u survey -p xxxx  mongodb://db-dev-706783.company.com/survey
MongoDB shell version v3.4.6
connecting to: mongodb://db-dev-706783.company.com/survey
MongoDB server version: 3.4.4
develop_RS:SECONDARY> 
bye

~$ mongo -u survey -p xxxx  mongodb://db-dev-706784.company.com/survey
MongoDB shell version v3.4.6
connecting to: mongodb://db-dev-706784.company.com/survey
MongoDB server version: 3.4.4
develop_RS:PRIMARY> 
bye

~$ mongo -u survey -p xxxx  mongodb://db-dev-706875.company.com/survey
MongoDB shell version v3.4.6
connecting to: mongodb://db-dev-706875.company.com/survey
MongoDB server version: 3.4.4
2017-07-25T10:19:27.178+0200 E QUERY    [thread1] Error: Authentication failed. :
DB.prototype._authOrThrow@src/mongo/shell/db.js:1461:20
@(auth):6:1
@(auth):1:2
exception: login failed
```

As far as we know it is not possible to auth against an arbiter node.

https://docs.mongodb.com/manual/core/replica-set-arbiter/
> Because arbiters do not store data, they do not possess the internal table of user and role mappings used for authentication. Thus, the only way to log on to an arbiter with authorization active is to use the localhost exception.

**Comment by ankhers** [comment]
Good news, @evax has given me permission to use their cluster to debug this issue. I will make a note here once I have found and fixed the issue.

**Comment by ankhers** [comment]
Would anyone having this issue please check #150 to see if it fixes your problem?

**Comment by ankhers** [comment]
I just merged #150. Please feel free to reopen this if you are still having issues with master.
